File: frontend/src/api/conversation.ts
import apiClient from './client';
import type {
  ApiResponse,
  Conversation,
  CreateConversationParams,
  Message,
  PageResponse,
} from '../types';

/**
 * 获取会话列表（分页）
 *
 * @param page 页码（从1开始）
 * @param size 每页数量
 * @returns 会话列表和是否有更多
 */
export async function getConversations(
  page: number = 1,
  size: number = 20
): Promise<{ items: Conversation[]; hasMore: boolean; total: number }> {
  const response = await apiClient.get<ApiResponse<PageResponse<Conversation>>>(
    '/conversation/list',
    { params: { page, size } }
  );
  const pageData = response.data.data;
  // 转换为兼容旧格式，同时保留 total
  return {
    items: pageData?.records || [],
    hasMore: (pageData?.current || 1) < (pageData?.pages || 0),
    total: pageData?.total || 0,
  };
}

/**
 * 创建新会话
 *
 * @param params 创建参数
 * @returns 新创建的会话
 */
export async function createConversation(
  params: CreateConversationParams
): Promise<Conversation> {
  // 后端 AI 会话创建接口
  const response = await apiClient.post<ApiResponse<Conversation>>(
    '/conversation/create/assistant',
    params
  );
  return response.data.data;
}

/**
 * 删除会话
 *
 * @param id 会话ID
 */
export async function deleteConversation(id: string): Promise<void> {
  await apiClient.delete(`/conversation/${id}`);
}

/**
 * 获取会话历史消息（返回完整消息树）
 *
 * @param conversationId 会话ID
 * @returns 包含所有消息和当前选中消息ID的响应
 */
export async function getConversationHistory(
  conversationId: string
): Promise<{ messages: Message[]; currentMessageId: string | null }> {
  const response = await apiClient.get<ApiResponse<{ messages: Message[]; currentMessageId: string | null }>>(
    `/conversation/history`,
    { params: { conversationId } }
  );
  return response.data.data || { messages: [], currentMessageId: null };
}

/**
 * 更新会话标题
 *
 * @param id 会话ID
 * @param title 新标题
 */
export async function updateConversationTitle(
  id: string,
  title: string
): Promise<void> {
  await apiClient.patch(`/conversation/modify`, { id, title });
}

File: backend/app/api/routes/conversation.py
from fastapi import APIRouter, Depends
from loguru import logger
from sqlalchemy.ext.asyncio import AsyncSession

from app.converter.converter import (
    ConversationConverter,
    MessageConverter,
)
from app.core.db import get_db_session
from app.dependencies.auth import CurrentUser, get_current_user
from app.schema.base import ApiResult
from app.schema.conversation import (
    ConversationParam,
    ConversationVo,
    HistoryResponse,
    MessageSendParam,
    MessageVo,
)
from app.services.conversation_service import ConversationService

router = APIRouter(prefix="/conversation", tags=["会话"])


@router.post("/create/assistant", response_model=ApiResult[ConversationVo])
async def create_assistant_conversation(
    payload: ConversationParam,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[ConversationVo]:
    """
    创建机器人会话，标题缺省时使用默认文案。
    """
    service = ConversationService(db)
    conv_id = await service.create_conversation(current.id, payload.title, payload.modelCode)
    conversation = await service.get_conversation(conv_id, current.id)
    return ApiResult.ok(ConversationConverter.from_dict(conversation))


@router.get("/list", response_model=ApiResult[dict])
async def list_conversations(
    limit: int = 20,
    offset: int = 0,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[dict]:
    """
    查询当前用户会话列表，支持分页。
    """
    service = ConversationService(db)
    items, has_more = await service.list_conversations(current.id, limit, offset)
    return ApiResult.ok(
        {
            "items": [ConversationConverter.from_dict(item) for item in items],
            "hasMore": has_more,
        }
    )


@router.get("/history", response_model=ApiResult[HistoryResponse])
async def history(
    conversationId: str,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[HistoryResponse]:
    """
    查询会话历史消息（返回完整消息树）
    """
    conv_service = ConversationService(db)
    # ForbiddenError 由全局异常处理器捕获
    await conv_service.ensure_owner(int(conversationId), current.id)
    result = await conv_service.history(current.id, int(conversationId))
    logger.info(f"[history] messages count: {len(result['messages'])}")
    return ApiResult.ok(
        HistoryResponse(
            messages=[MessageConverter.from_dict(item) for item in result["messages"]],
            currentMessageId=result["currentMessageId"],
        )
    )


@router.get("/{conversation_id}", response_model=ApiResult[ConversationVo])
async def get_conversation_detail(
    conversation_id: str,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[ConversationVo]:
    """
    获取会话详情，校验归属。
    """
    service = ConversationService(db)
    vo = await service.get_conversation(int(conversation_id), current.id)
    return ApiResult.ok(ConversationConverter.from_dict(vo))


@router.post("/send", response_model=ApiResult[MessageVo])
async def send_message(
    payload: MessageSendParam,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[MessageVo]:
    """
    同步消息发送并落库。
    """
    service = ConversationService(db)
    await service.ensure_owner(int(payload.conversationId), current.id)
    message = await service.persist_message(
        conversation_id=int(payload.conversationId),
        sender_id=current.id,
        role="user",
        content=payload.content,
        content_type=payload.contentType or "TEXT",
    )
    return ApiResult.ok(MessageConverter.to_vo(message))


# /history 路由已移到 /{conversation_id} 之前定义


@router.patch("/modify", response_model=ApiResult[None])
async def modify_conversation(
    payload: ConversationParam,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[None]:
    """
    修改会话标题。
    """
    service = ConversationService(db)
    await service.modify_conversation(
        current.id, int(payload.id) if payload.id else None, payload.title
    )
    return ApiResult.ok()


@router.delete("/{conversation_id}", response_model=ApiResult[None])
async def delete_conversation(
    conversation_id: str,
    db: AsyncSession = Depends(get_db_session),
    current: CurrentUser = Depends(get_current_user),
) -> ApiResult[None]:
    """
    删除会话及其消息。
    """
    service = ConversationService(db)
    await service.delete_conversation(current.id, int(conversation_id))
    return ApiResult.ok()
